apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/DecimalReadTests.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright 2022 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Spanner.Data.IntegrationTests
+{
+    [Collection(nameof(DecimalTableFixture))]
+    public class DecimalReadTests
+    {
+        private readonly DecimalTableFixture _fixture;
+
+        public DecimalReadTests(DecimalTableFixture fixture) =>
+            _fixture = fixture;
+
+        public static IEnumerable<object[]> TestDecimals =>
+            new List<object[]>
+            {
+                new object[] { new decimal(1.23456789) },
+                new object[] { decimal.MaxValue },
+                new object[] { decimal.MinValue },
+                new object[] { decimal.MinusOne },
+                new object[] { decimal.Zero },
+                new object[] { new decimal(123456789.0123456789012345678) },
+            };
+
+        [Theory]
+        [MemberData(nameof(TestDecimals))]
+        public async Task WriteThenRead_ShouldBeEqual(decimal expectedValue)
+        {
+            using var connection = _fixture.GetConnection();
+            // Write the decimal value and read it back.
+            await connection.CreateInsertOrUpdateCommand(_fixture.TableName,
+                new SpannerParameterCollection
+                {
+                        new SpannerParameter("DecimalValue", SpannerDbType.Float64, expectedValue),
+                }
+            ).ExecuteNonQueryAsync();
+            var decimalValue = await connection.CreateSelectCommand($"SELECT DecimalValue FROM {_fixture.TableName}").ExecuteScalarAsync<decimal>();
+            Assert.Equal(expectedValue, decimalValue);
```

Comment:
I wouldn't be surprised by this code failing - I'd expect it to lose precision, given that we're trying to store more than 64 bits of significant data in a FLOAT64 type.

I'll pull the PR to see what actually happens though...

Reply:
Okay, this isn't actually a Spanner issue really - it's just lossy conversions that then overflow. This throws an exception too:

```csharp
using System;

decimal x = decimal.MaxValue;
double d = (double) x;
decimal y = (decimal) d;
```

Basically the conversion to `double` is rounding up, at which point we're out of the range of `decimal`.